### PR TITLE
API model stubs

### PR DESCRIPTION
**Pull request by katstevens**
Trello: https://trello.com/c/FYPvix7N/270-update-api-stubs-in-dm-utils

Intended as a replacement for `dmutils.api_stubs`, which was outdated, inconsistent and incomplete (see https://github.com/alphagov/digitalmarketplace-utils/blob/master/dmutils/api_stubs.py). 

- Stubs for Brief, Supplier, Lot, Framework, FrameworkAgreement, SupplierFramework, AuditEvent
- Two methods for each stub class, `response()` and `single_result_response()` which return the object serialization and the API response (i.e. including the top level key) respectively. Previously the stubs were a mixture of these two response formats 😿 
- I've kept the `_dos_lots` and `_cloud_lots` functions as they're a useful shorthand for generating DOS and G-Cloud frameworks.
- I've also kept the ability to provide framework date kwargs as a `datetime` object, converting them to datestamp strings. However to avoid requiring `dmutils` as a dependency, I've hardcoded the format instead of importing `DATETIME_FORMAT`.

I've tried to keep backwards incompatible arguments where possible, but the `ModelnameStub.response()` format is going to need updating everywhere anyway, so this will not be particularly seamless to integrate. 

To avoid these going out of date again, I intend to write tests in the API that compare the stub keys with the model serialization keys. That way, when a new field gets added, it'll be immediately obvious that these stub needs updating.

### CODE REVIEW

**Comment by lfdebrux** [comment]
I love where this is going, awesome work 🥇 

I don't know how other people feel about this, but I feel that it might benefit from being a folder with multiple files rather than one big file; what are your thoughts?

File: dmtestutils/api_model_stubs.py [hunk 1]
```diff
@@ -0,0 +1,509 @@
...
+            self.response_data['users'] = [self.user]
+
+        # Allow snake case framework_* kwargs for backwards compatibility
+        for nested_framework_key, camelcase_key, snakecase_kwarg in [
+            ('family', 'frameworkFramework', 'framework_family'),
+            ('name', 'frameworkName', 'framework_name'),
+            ('slug', 'frameworkSlug', 'framework_slug'),
+            ('status', 'frameworkStatus', 'framework_status'),
+        ]:
+            if kwargs.get(snakecase_kwarg):
```
[the remark this answers is not in the document]
Comment (katstevens) [reply dmtestutils/api_model_stubs.py]: I'm pretty sure we don't use it as a kwarg anywhere - `frameworkFramework` is deprecated, we should be using 'family' everywhere instead. I've kept it in the serialization as it's still included in the API response.

File: dmtestutils/api_model_stubs.py [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+    resource_name = 'briefResponses'
+    default_data = {
+
+    }
+
+
+class DirectAwardProjectStub(BaseAPIModelStub):
+    resource_name = 'project'
+    default_data = {
+
```
[the remark this answers is not in the document]
Comment (katstevens) [reply dmtestutils/api_model_stubs.py]: Ha probably! I'll add a note in the docstring that the other stubs need expanding.

File: dmtestutils/api_model_stubs.py [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+class BaseAPIModelStub:
+
+    resource_name = None
+    default_data = {}
+    optional_keys = []
+
+    def __init__(self, **kwargs):
+        self.response_data = self.default_data.copy()
+
+        # Backwards compatibility for snake case kwargs
```
[the remark this answers is not in the document]
Comment (katstevens) [reply dmtestutils/api_model_stubs.py]: I'd rather keep dependencies on testutils to a minimum - I think it's useful to have some visibility on what sort of snake case kwargs we're allowing in the stubs anyway.

[the remark this answers is not in the document]
Comment (lfdebrux) [reply dmtestutils/api_model_stubs.py]: 👍

**Comment by katstevens** [comment]
I could be convinced about splitting up into files - were you thinking one file per stub? Or grouping them somehow? I assume we'd still import everything into the `__init__` anyway to avoid clunkiness in the actual tests?

**Comment by lfdebrux** [comment]
> I could be convinced about splitting up into files - were you thinking one file per stub? Or grouping them somehow? I assume we'd still import everything into the `__init__` anyway to avoid clunkiness in the actual tests?

I was thinking one file per stub. I agree about the `__init__`.

**Comment by katstevens** [comment]
@lfdebrux Now reorganised! Thanks for the suggestion, I agree it looks better. I kept the tests all in one module though as having complex tests on our test utils is an inception step to far 😹.

I've checked they still work for my Supplier FE and Admin FE branches.

**Pull request merged**